### PR TITLE
enh(export conf): Updated export files to allow adding export of plugin commands in PP manager

### PR DESCRIPTION
**Pull request by callapa**
## Description

Updated export files to allow adding export of plugin commands in PP manager

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/class/config-generate/plugins.class.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ */
+
+declare(strict_types=1);
+
+require_once 'abstract/JsonWriter.class.php';
+
+/**
+ * This class is designed to export commands from pluginpacks.
+ */
+class Plugins extends JsonWriter
```
Comment (kduret) [inline www/class/config-generate/plugins.class.php:44]: is it possible to move it in centreon-pp-manager repository ?
Like it's done for business activity generation : https://github.com/centreon/centreon-bam/tree/develop/www/modules/centreon-bam-server/generate_files

File: www/class/config-generate/abstract/JsonFormat.class.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+declare(strict_types=1);
+
+abstract class JsonFormat
+{
+    /**
+     * @var mixed
+     */
+    protected mixed $cacheData = null;
+
+    protected ?string $filePath = null;
```
Comment (kduret) [inline www/class/config-generate/abstract/JsonFormat.class.php:46]:
```suggestion
    /**
     * @var string|null
     */
    protected ?string $filePath = null;
```

**Review by kduret — DISMISSED** [review]

File: www/class/config-generate/SingletonTrait.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * @param \Pimple\Container $dependencyInjector
+     * @return static
+     */
+    public static function getInstance(\Pimple\Container $dependencyInjector): static
+    {
+        /**
+         * @var array<string, static>
+         */
+        static $instances = array();
```
Comment (kduret) [inline www/class/config-generate/SingletonTrait.php:50]:
```suggestion
        static $instances = [];
```

File: www/class/config-generate/abstract/JsonFormat.class.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        $this->filePath = $filePath;
+    }
+
+    /**
+     * Writes the content of the cache only if it is not empty.
+     *
+     * @return int|bool Number of bytes written or false on failure
+     * @throws Exception
+     */
+    public function flushContent(): int|bool
```
Comment (kduret) [inline www/class/config-generate/abstract/JsonFormat.class.php:75]:
```suggestion
    public function flushContent(): void
```

File: www/class/config-generate/abstract/JsonFormat.class.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        if ($this->filePath === null) {
+            throw new Exception('No file path defined');
+        }
+        if (! empty($this->cacheData)) {
+            $data = json_encode($this->cacheData, JSON_UNESCAPED_SLASHES | JSON_PRETTY_PRINT);
+            $this->cacheData = null;
+
+            return file_put_contents($this->filePath, $data);
```
Comment (kduret) [inline www/class/config-generate/abstract/JsonFormat.class.php:84]: Could you throw an exception on failure

File: www/class/config-generate/command.class.php [hunk 6]
```diff
@@ -117,4 +121,23 @@ public function generateFromCommandId($command_id)
...
     }
+
+    /**
+     * Get information of command.
+     *
+     * @param int $commandId
+     * @return array{
+     *     command_name: string,
+     *     command_line_base: string,
+     *     connector: string,
```
Comment (kduret) [inline www/class/config-generate/command.class.php:132]:
```suggestion
     *     connector: string|null,
```

File: www/class/config-generate/generate.class.php [hunk 7]
```diff
@@ -344,12 +345,16 @@ public function generateModuleObjects($type = 1): void
...
         if (is_array($this->module_objects)) {
             foreach ($this->module_objects as $module_object) {
+                $externalModule = $module_object::getInstance($this->dependencyInjector);
                 if (
-                    ($type == 1 && $module_object::getInstance($this->dependencyInjector)->isEngineObject() == true)
-                    || ($type == 2 && $module_object::getInstance($this->dependencyInjector)->isBrokerObject() == true)
+                    $externalModule instanceof ExternalModuleGenerationInterface
+                    && (
+                        ($type == 1 && $externalModule->isEngineObject() === true)
+                        || ($type == 2 && $externalModule->isBrokerObject() == true)
```
Comment (kduret) [inline www/class/config-generate/generate.class.php:353]:
```suggestion
                        || ($type == 2 && $externalModule->isBrokerObject() === true)
```

**Review by a-launois — DISMISSED** [review]

**Review by kduret — DISMISSED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

**Review by kduret — DISMISSED** [review]

**Review by a-launois — DISMISSED** [review]

**Review by kduret — DISMISSED** [review]

**Review by a-launois — DISMISSED** [review]

**Review by kduret — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**